### ISSUE TITLE
Begin using the "override" and "final" specifiers.

### ISSUE THREAD

**Issue by bakercp** [openframeworks/openFrameworks#4856]
In our documentation discussions (and based on the recent survey), it's clear that users find answers to their questions using a multiplicity of strategies.  Many advanced users read the headers directly for answers.  Thus, our push for "const-correctness" in recent years has made it much easier to determine essential qualities of the interface (e.g. passing pixels to function A will not modify the passed pixels) without digging too far into the implementation.  Of course it also goes far to help users avoid mistakes.

Since our codebase heavily relies upon inheritance and virtual hierarchies that can sometimes be confusing requiring users to drill down into our powerful virtual inheritance hierarchy (especially around pixels and video).  

With the new support of c++11, I'd propose that continue building code that is clearer and easier to read for advanced users by employing the consistent use of the new `override` and `final` specifiers (http://en.cppreference.com/w/cpp/language/override and http://en.cppreference.com/w/cpp/language/final).  

While I consider these specifiers to be valuable from a programming perspective (e.g. the application of `final` could be useful in some cases), they clearly signal and document intent API design intent, which in my mind works in concert with the written low-level doxygen-style documentation.  

Timing on implementing this is clearly a concern --  a consistent application would touch a lot of headers -- but perhaps we could pick a milestone and move in that direction.

Like the successful const-correct push, it would not change functionality.

Feedback?

@arturoc @workergnome @ofZach @danomatika

**Comment by arturoc** [comment]
i'd disagree that we rely "heavily on inheritance" there's some base classes that are only providing polymorphism, they are pure abstract , equivalent to interfaces in other languages and don't have any implementation so there's really no inheritance in the traditional sense that some classes "inherit" behaviour or even data from others, just polymorphism.

I've been specially careful to never rely on inheritance tried to not have implementation and mostly data in the base classes.

There's only 1 case i can think of where inheritance is really used, ofNode and everything that inherits from it.

being that the case i think override and final are not that important at least not for the base classes and anything that implements them. having used something similar to what will be concepts in c++ i think it'll probably be a better choice for what we have now as base classes and in that case we would have to revert this keywords everywhere since they don't apply in the case of concepts.

I'm not pretty sure on the details about how concepts will work yet but i don't think either that using this keywords is really important given that the base classes are purely abstract so i would wait to add them.

**Comment by bakercp** [comment]
Maybe a clarification of inheritance would be helpful -- 

For ofVideoPlayer we use something that might better be described as a SPIMPL (smart pointer to implementation).  I guess I'm referring to this kind of inheritance ...

![openframeworks__ofavfoundationplayer_class_reference](https://cloud.githubusercontent.com/assets/300484/13180289/0b79e088-d6e5-11e5-86fa-48114d5f73a0.png)

**Comment by arturoc** [comment]
yes, that's what i meant there's really only a way to provide polymorphism for the video implementers there, ofAVFoundationPlayer implements some base abstract classes but those don't provide any behaviour or data.

From wikipedia:

> Inheritance should not be confused with subtyping.[2][3] In some languages inheritance and subtyping agree,[a] whereas in others they differ; in general subtyping establishes an is-a relationship, whereas inheritance only reuses implementation and establishes a syntactic relationship, not necessarily a semantic relationship (inheritance does not ensure behavioral subtyping). To distinguish these concepts, subtyping is also known as interface inheritance, whereas inheritance as defined here is known as implementation inheritance or code inheritance.[4] Still, inheritance is a commonly used mechanism for establishing subtype relationships.[5]

What we have here is just subtyping, there's no specific keyword in c++ to diferentiate but since all the base classes are pure abstract is the implementers are just subtypes, there's nothing to inherit, no implementation or data in the base classes.

in the relation from ofNode to ofLight, ofCamera... there's inheritance cause the extending classes inherit the behaviour from ofNode

**Comment by bakercp** [comment]
Got it.  I'll clarify the language above. The `concepts` approach is definitely interesting though.  Can you recommend some reading / references 
on what that might look like?

With the interface inheritance language clarified, I'd still suggest that we should consider these specifiers where appropriate.

**Comment by bakercp** [comment]
Are we against using `override`, `final`, to make the code clearer (regardless of what you call inheritance / concepts, etc)?  I believe it's an opportunity for clearer code, just like using `const`, etc. @arturoc @ofZach @ofTheo
